pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,360 @@
+
+    Applying 10 steps gets us close the ground state of :math:`E\approx-2.23`
+
+    >>> for step in range(10):
+    ...    cost = opt.step_and_cost()
+    ...    print(f"Step {step} - cost {cost}")
+    Step 0 - cost -1.3351865007304005
+    Step 1 - cost -1.9937887238935206
+    Step 2 - cost -2.1524234485729834
+    Step 3 - cost -2.1955105378898487
+    Step 4 - cost -2.2137628169764256
+    Step 5 - cost -2.2234364822091575
+    ...
+
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, circuit, stepsize=0.01, restriction=None, exact=False, trottersteps=1):
+
+        if not isinstance(circuit, qml.QNode):
+            raise TypeError(f"circuit must be a QNode, received {type(circuit)}")
+
+        self.circuit = circuit
+        self.circuit.construct([], {})
+        self.hamiltonian = circuit.func().obs
+        if not isinstance(self.hamiltonian, qml.Hamiltonian):
+            raise TypeError(
+                f"circuit must return the expectation value of a Hamiltonian,"
+                f"received {type(circuit.func().obs)}"
+            )
+        self.nqubits = len(circuit.device.wires)
+
+        if self.nqubits > 4:
+            warnings.warn(
+                "The exact Lie gradient is exponentially expensive in the number of qubits,"
+                f"optimizing a {self.nqubits} qubit circuit may be slow.",
+                UserWarning,
+            )
+        if restriction is not None and not isinstance(restriction, qml.Hamiltonian):
+            raise TypeError(
+                f"restriction must be a Hamiltonian, received {type(restriction)}"
+            )
+        (
+            self.lie_algebra_basis_ops,
+            self.lie_algebra_basis_names,
+        ) = self.get_su_n_operators(restriction)
+        self.exact = exact
+        self.trottersteps = trottersteps
+        self.coeffs, self.observables = self.hamiltonian.terms
+        self.stepsize = stepsize
+
+    def step_and_cost(self):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Returns:
+             float: the objective function output prior to the step.
+
+        """
+        cost = self.circuit()
+        omegas = self.get_omegas()
+        nonzero_idx = np.where(~np.isclose(omegas, 0))[0]
+
+        non_zero_lie_algebra_elements = [self.lie_algebra_basis_names[i] for i in nonzero_idx]
+        non_zero_omegas = [-omegas[i] for i in nonzero_idx]
+
+        lie_gradient = qml.Hamiltonian(
+            non_zero_omegas,
+            [
+                qml.grouping.string_to_pauli_word(ps)
+                for ps in non_zero_lie_algebra_elements
+            ],
+        )
+        new_circuit = append_time_evolution(lie_gradient, self.stepsize, self.trottersteps, self.exact)(
+            self.circuit.func
+        )
+
+        self.circuit = qml.QNode(new_circuit, self.circuit.device)
```

Comment:
just double checking, but it is intentional to build the new circuit here, but not evaluate it?

Reply:
I had it the other way around before, but similar to the other optimizers we calculate the cost before the 'step'. So it would get evaluated on the next step. Maybe this could give weird behavior if you only let the optimizer take one step?

Reply:
Ah, the only reason the other optimizers work like that is that it is taking advantage of the fact that, when using autodiff frameworks, a forward pass always occurs before the 'backwards' pass - so when you compute the gradient, you get the cost 'for free'.

If you think it makes more sense (both from a user perspective and a logical perspective) to compute the cost _post_ step, feel free to do so. Just as long as the behaviour is clearly documented in the `step_and_cost` docstring!

Reply:
I think it would make sense to calculate stuff pre-optimization so that it's the same for all optimizers.